Added biome color cache support for custom color resolver

CPTProgrammer opened the pull request:
Added biome color cache support for custom color resolver. 

`WorldSlice#getColor` can now be invoked with a custom `ColorResolver`.

- Fixed #2221
- Fixed #2166

CPTProgrammer committed:
Added biome color cache support for custom color resolver

jellysquid3 commented on src/main/java/me/jellysquid/mods/sodium/client/world/biome/BiomeColorCache.java:
```diff
@@ -94,6 +137,29 @@ private void updateColorBuffers(int relY) {
+                int relativeX = worldX - this.minX;
+                int relativeZ = worldZ - this.minZ;
+
+                slice.buffer.set(relativeX, relativeZ, BiomeColors.GRASS_COLOR.getColor(biome, worldX, worldZ));
```
This does not actually use the resolver passed to the function, so custom resolvers will always use the default grass color...

CPTProgrammer replied:
it has been fixed :)

CPTProgrammer commented:
Sorry for my carelessness, I'll change the code as soon as possible

CPTProgrammer committed:
Update BiomeColorCache.java

Merge "custom" and "vanilla" color resolvers.

Fixed issue where custom resolvers would always use the default grass color.

jellysquid3 commented on src/main/java/me/jellysquid/mods/sodium/client/world/biome/BiomeColorCache.java:
```diff
@@ -6,15 +6,17 @@
-    private final Slice[] slices;
-    private final boolean[] populatedSlices;
+    private final HashMap<ColorResolver, Slice[]> slices;
+    private final HashMap<ColorResolver, boolean[]> populatedSlices;
```
We should avoid having two hash maps, and instead just store the "is populated" flag inside the `Slice` since it avoids the overhead of additional lookups, and makes the code cleaner.

jellysquid3 commented on src/main/java/me/jellysquid/mods/sodium/client/world/biome/BiomeColorCache.java:
```diff
@@ -23,21 +25,19 @@ public class BiomeColorCache {
-        for (int y = 0; y < sizeY; y++) {
-            this.slices[y] = new Slice(sizeXZ);
-        }
+        this.slices = new HashMap<>();
```
Should be using `Reference2ReferenceOpenHashMap` since performance is critical, and caches are only valid according to reference equality. The type of the field should also be updated to match.

jellysquid3 commented on src/main/java/me/jellysquid/mods/sodium/client/world/biome/BiomeColorCache.java:
```diff
@@ -49,26 +49,47 @@ public void update(ChunkRenderContext context) {
-        if (!this.populatedSlices[relY]) {
-            this.updateColorBuffers(relY);
+        if (!this.slices.containsKey(resolver)) {
+            this.slices.put(resolver, new Slice[this.sizeY]);
```
This branch should be extracted to a new function since it's rather large and a rarely taken path (we don't want the CPU fetching instructions into the cache which are not helpful in most invocations).

jellysquid3 commented on src/main/java/me/jellysquid/mods/sodium/client/world/biome/BiomeColorCache.java:
```diff
@@ -49,26 +49,47 @@ public void update(ChunkRenderContext context) {
-    private void updateColorBuffers(int relY) {
-        var slice = this.slices[relY];
+    private void updateColorBuffers(int relY, ColorResolver resolver) {
+        var slice = this.slices.get(resolver)[relY];
```
We already created the slice before calling this method, so we should pass it to this function instead of querying it again.

jellysquid3 reviewed (changes requested):
The abstract of this looks good, just some performance concerns.

CPTProgrammer commented on src/main/java/me/jellysquid/mods/sodium/client/world/biome/BiomeColorCache.java:
```diff
@@ -49,26 +49,47 @@ public void update(ChunkRenderContext context) {
-        Arrays.fill(this.populatedSlices, false);
+        for (boolean[] p : this.populatedSlices.values()) {
+            Arrays.fill(p, false);
+        }
```
Optimizing performance by using an incremental flag (without loops) to check for populated status instead of relying on two nested for-loops.

CPTProgrammer reviewed (commented):
Performance optimization

CPTProgrammer committed:
Performance optimization for BiomeColorCache

- Utilize Reference2ReferenceOpenHashMap: Replaced the usage of `HashMap` with `Reference2ReferenceOpenHashMap`.
- Slice Initialization Refactoring: Extracted the slice initialization branch into a new function called `initializeSlices`.
- Enhanced Populated Status Check: Implemented an incremental flag `populateStamp` (without loops) to check for populated status, replacing the reliance on two nested for-loops.

CPTProgrammer commented:
All requested changes have been implemented

embeddedt reviewed (commented):
Looks good from my end as well.

MehVahdJukaar commented:
Will this be added to previous version aswell? 
The fact that its not possible to add custom ColorResolvers is a deal breakers for many mods that use those things and its made way worse by how locked and unforgiving up the current architecture is with that enum that cant be extended or the fact that the game hardcrashes when a color resolved that is not vanilla shows up.
This is relevant to my mod [Polytone](https://legacy.curseforge.com/minecraft/mc-mods/polytone) which currently crashes with Sodium
